benchmark/opperf/utils/op_registry_utils.py

```diff
@@ -0,0 +1,193 @@
+    Contrib operators that starts with '_contrib' and internal operators that
+    starts with '_'.
+
+    Parameters
+    ----------
+    filters: tuple(str)
+        List of operator name prefix to ignore from benchmarking.
+        Default - ("_backward", "_contrib", "_")
+
+    Returns
+    -------
+    {"operator_name": {"has_backward", "nd_op_handle"}}
+    """
+    plist = ctypes.POINTER(ctypes.c_char_p)()
+    size = ctypes.c_uint()
+
+    check_call(_LIB.MXListAllOpNames(ctypes.byref(size),
+                                     ctypes.byref(plist)))
+    mx_operators = {}
+    operators_with_backward = []
+
+    # Prepare master list of all operators.
+    for i in range(size.value):
+        cur_op_name = py_str(plist[i])
+        if not cur_op_name.startswith(filters):
+            mx_operators[cur_op_name] = {"has_backward": False,
+                                         "nd_op_handle": getattr(mx_nd_module, cur_op_name)}
+
+        if cur_op_name.startswith("_backward_"):
+            operators_with_backward.append(cur_op_name)
+
+    # Identify all operators that can run backward.
+    for op_with_backward in operators_with_backward:
+        op_name = op_with_backward.split("_backward_")[1]
+        if op_name in mx_operators:
+            mx_operators[op_name]["has_backward"] = True
+
+    return mx_operators
+
+
+def _get_op_handles(op_name):
+    """Get handle for an operator with given name - op_name.
+
+    Parameters
+    ----------
+    op_name: str
+        Name of operator to get handle for.
+    """
+    op_handle = OpHandle()
+    check_call(_LIB.NNGetOpHandle(c_str(op_name), ctypes.byref(op_handle)))
+    return op_handle
+
+
+def _get_op_arguments(op_name, op_handle):
+    """Given operator name and handle, fetch operator arguments - number of arguments,
+    argument names, argument types.
+
+    Parameters
+    ----------
+    op_name: str
+        Name of the operator
+
+    op_handle: OpHandle
+        Handle for the operator
+
+    Returns
+    -------
+    (narg, arg_names, arg_types)
+    """
+    real_name = ctypes.c_char_p()
+    desc = ctypes.c_char_p()
+    num_args = mx_uint()
+    arg_names = ctypes.POINTER(ctypes.c_char_p)()
+    arg_types = ctypes.POINTER(ctypes.c_char_p)()
+    arg_descs = ctypes.POINTER(ctypes.c_char_p)()
+    key_var_num_args = ctypes.c_char_p()
+    ret_type = ctypes.c_char_p()
+
+    check_call(_LIB.MXSymbolGetAtomicSymbolInfo(
+        op_handle, ctypes.byref(real_name), ctypes.byref(desc),
+        ctypes.byref(num_args),
+        ctypes.byref(arg_names),
+        ctypes.byref(arg_types),
+        ctypes.byref(arg_descs),
+        ctypes.byref(key_var_num_args),
+        ctypes.byref(ret_type)))
+
+    narg = int(num_args.value)
+    arg_names = [py_str(arg_names[i]) for i in range(narg)]
+    arg_types = [py_str(arg_types[i]) for i in range(narg)]
+
+    return narg, arg_names, arg_types
+
+
+def _set_op_arguments(mx_operators):
```

Comment:
I would put op assembling together in a single function and use a light class or named tuple instead of difficult to maintain deeply nested dictionaries.

Reply:
Getting a op handle by op name, getting a op arguments by op handle are all independent unit of work hence kept it separate. This assembling function just gets op handle, op argument and prepares a dictionary. I agree we could have a class / named tuple. But, to keep the utility simple, flexible and hackable kept it this way. For now, we have only 3 attributes, name, has_backward, params. I prefer to keep it this way for now and revisit this later if necessary. Let me know if you still strongly think otherwise.